File: 2022-04/python/542_01_Matrix.py
from collections import deque

class Solution:

    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        queue = deque()
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                if mat[i][j] == 0:
                    queue.append((i, j))
                else:
                    mat[i][j] = -1
        while queue:
            i, j = queue.popleft()
            for x, y in [(i + 1, j), (i - 1, j), (i, j - 1), (i, j + 1)]:
                if 0 <= x < len(mat) and 0 <= y < len(mat[0]) and mat[x][y] == -1:
                    mat[x][y] = mat[i][j] + 1
                    queue.append((x, y))
        return mat

File: 2022-04/javascript/542_01_Matrix.js
/**
 * @param {number[][]} mat 
 * @return {number[][]} 
 */
function updateMatrix(mat) {
    const queue = []
    for (let i = 0; i < mat.length; i++) {
        for (let j = 0; j < mat[0].length; j++) {
            if (mat[i][j] === 0) {
                queue.push({i, j})
            } else {
                mat[i][j] = -1
            }
        }
    }
    while (queue.length > 0) {
        const {i, j} = queue.shift()
        const dirs = [
            {x: i + 1, y: j},
            {x: i - 1, y: j},
            {x: i, y: j - 1},
            {x: i, y: j + 1},
        ]
        for (let {x, y} of dirs) {
            if (x >= 0 && x < mat.length && y >= 0 && y < mat[0].length && mat[x][y] === -1) {
                mat[x][y] = mat[i][j] + 1
                queue.push({i: x, j: y})
            }
        }
    }
    return mat
}